File: DSA_Level-1/Basics Of Programming/Function And Arrays/First Index And Last Index/FirstIndexAndlastIndex_Solution.js

function firstIndex(nums,d){
    let low=0,high=nums.length-1;
    let mid,res=-1;;
    while(low<=high){
        mid=parseInt((low+high)/2);
        if(nums[mid]>d){
            high=mid-1;
        }
        else if(nums[mid]<d){
            low=mid+1;
        }
        else{
            high=mid-1;
            res=mid;
        }
    }

    return res;
}

function lastIndex(nums,d){
    let low=0,high=nums.length-1;
    let mid,res=-1;;
    while(low<=high){
        mid=parseInt((low+high)/2);
        if(nums[mid]>d){
            high=mid-1;
        }
        else if(nums[mid]<d){
            low=mid+1;
        }
        else{
            low=mid+1;
            res=mid;
        }
    }

    return res;
}


function firstAndLastIndex(nums,d){
    let firstIdx=firstIndex(nums,d);
    let lastIdx=lastIndex(nums,d);

    return {firstIdx,lastIdx};
}

console.log(firstAndLastIndex([1,2,3,3,3,3,3,3,4,5,6,7,8,9],3));

File: DSA_Level-1/Basics Of Programming/Function And Arrays/First Index And Last Index/FirstIndexAndLastIndex_Solution.java

import java.io.*;
import java.util.*;

public class Main{

public static void main(String[] args) throws Exception {
    Scanner scn=new Scanner(System.in);
    int n=scn.nextInt();
    int[] nums=new int[n];
    for(int i=0;i<n;++i){
        nums[i]=scn.nextInt();
    }
    
    int d=scn.nextInt();
    
    int firstIdx=firstIndex(nums,d);
    int lastIdx=lastIndex(nums,d);
    
    System.out.println(firstIdx);
    System.out.println(lastIdx);
 }
 
 public static int firstIndex(int[] nums,int d){
    int low=0,high=nums.length-1;
    int mid,res=-1;;
    while(low<=high){
        mid=((low+high)/2);
        if(nums[mid]>d){
            high=mid-1;
        }
        else if(nums[mid]<d){
            low=mid+1;
        }
        else{
            high=mid-1;
            res=mid;
        }
    }

    return res;
}

public static int lastIndex(int[] nums,int d){
    int low=0,high=nums.length-1;
    int mid,res=-1;;
    while(low<=high){
        mid=((low+high)/2);
        if(nums[mid]>d){
            high=mid-1;
        }
        else if(nums[mid]<d){
            low=mid+1;
        }
        else{
            low=mid+1;
            res=mid;
        }
    }

    return res;
}

}